chemprop/models/ffn.py

```diff
@@ -0,0 +1,439 @@
+            b_hidden_2 = torch.cat([backward_bond, forward_bond], dim=1)
+            output_1 = self.ffn_readout(b_hidden_1)
+            output_2 = self.ffn_readout(b_hidden_2)
+            output = (output_1 + output_2) / 2
+            if bond_types is not None:
+                output = output + bond_types.reshape(-1, 1)
+
+        return output
+
+class FFNAtten(nn.Module):
+    """
+    A :class:`FFNAtten` is a multiple feed forward neural networks (FFN) to predict
+    the atom/bond descriptors with applying constraint on output. An attention-based
+    constraint is used. This method is from `Regio-selectivity prediction with a
+    machine-learned reaction representation and on-the-fly quantum mechanical descriptors
+    <https://pubs.rsc.org/en/content/articlelanding/2021/sc/d0sc04823b>`_, section 2.2.
+    """
+
+    def __init__(
+        self,
+        features_size: int,
+        hidden_size: int,
+        num_layers: int,
+        output_size: int,
+        dropout: float,
+        activation: nn.Module,
+        ffn_base: nn.Module,
+        ffn_type: str = "atom",
+        shared_ffn: bool = True,
+        weights_ffn_num_layers: int = 2,
+    ):
+        """
+        :param features_size: Dimensionality of input features.
+        :param hidden_size: Dimensionality of hidden layers.
+        :param num_layers: Number of layers in FFN.
+        :param output_size: The size of output.
+        :param dropout: Dropout probability.
+        :param activation: Activation function.
+        :param ffn_base: The shared base layers (all but the last) of the FFN between tasks.
+        :param ffn_type: The type of target (atom or bond).
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super().__init__()
+
+        if num_layers == 1:
+            base_output_size = features_size
+        else:
+            base_output_size = hidden_size
+
+        if shared_ffn:
+            self.ffn = ffn_base
+        else:
+            if num_layers > 1:
+                self.ffn = nn.Sequential(
+                    build_ffn(
+                        first_linear_dim=features_size,
+                        hidden_size=hidden_size,
+                        num_layers=num_layers - 1,
+                        output_size=hidden_size,
+                        dropout=dropout,
+                        activation=activation,
+                    ),
+                    activation,
+                )
+            else:
+                self.ffn = nn.Identity()
+        self.ffn_readout = build_ffn(
+            first_linear_dim=base_output_size,
+            hidden_size=hidden_size,
+            num_layers=1,
+            output_size=output_size,
+            dropout=dropout,
+            activation=activation,
+        )
```

Comment:
this is just a linear layer, why build it using some abstract function?

Reply:
The reason I wrote it is for consistency (building every FFN via `build_ffn()` function), but this one is relatively simpler so I should replace this block with the below code.
```
self.ffn_readout = nn.Sequential(
    nn.Dropout(dropout),
    nn.Linear(base_output_size, output_size),
)
```